airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py

```diff
@@ -0,0 +1,142 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+import concurrent.futures
+import datetime
+import math
+from abc import ABC
+from dataclasses import asdict
+from typing import Any, Dict, Iterable, List, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+
+from .api import ZohoAPI
+from .exceptions import IncompleteMetaDataException, UnknownDataTypeException
+from .types import FieldMeta, ModuleMeta, ZohoPickListItem
+
+
+class ZohoCrmStream(HttpStream, ABC):
+    primary_key: str = "id"
+    module: ModuleMeta = None
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        if response.status_code != 200:
+            return None
+        pagination = response.json()["info"]
+        if not pagination["more_records"]:
+            return None
+        return {"page": pagination["page"] + 1}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return next_page_token or {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        data = response.json()["data"] if response.status_code == 200 else []
```

Comment:
Any non-obvious logic like this should have a comment helping the reader understand why we’re doing it. For example why do we skip on non-200 response? Shouldn’t we raise ?

Reply:
This was intended to handle statuses 204 and 304 which are successful response statuses with empty body in Zoho API. Changed this condition to be more narrow and introduced EMPTY_BODY_STATUSES variable to make it more obvious